thorlcr/slave/traceslave.hpp

```diff
@@ -0,0 +1,226 @@
+                rowTailp = 0;
+            }
+        }
+        inline void enqueue(const void *row, unsigned sequence, unsigned _msTick)
+        {
+            // NOTE - updates then sets rowHeadp (atomically) so that the code
+            // below (ignoring oldest record) is safe
+            rowBuffer[rowTailp].set(row,sequence,_msTick);
+            unsigned nextTailp = rowTailp + 1;
+            if (nextTailp==rowBufferSize)
+                nextTailp = 0;
+            rowTailp = nextTailp;
+        }
+        inline void enqueue(const CThorRowHistory &_rowBuffer)
+        {
+            enqueue(_rowBuffer.get(),_rowBuffer.getSequence(),_rowBuffer.getmsTickStamp());
+        }
+        // Move items in current queue into a second queue (clear out 'this' queue)
+        // NOTE - set skipLast to true when the oldest item must be skipped.  This should
+        // be used where the queue COULD have been updated after the buffers are swapped and
+        // the enqueueRowForTrace overwriting the oldest entry.
+        void queueOut(TraceQueue &traceOutQueue, bool skipOldestEntry)
+        {
+            const unsigned numberRowsToSkip = skipOldestEntry ? 1 : 0;
+            unsigned rowPos = rowTailp + numberRowsToSkip;
+            if (rowPos==rowBufferSize) rowPos = 0;
+            for (unsigned n=numberRowsToSkip; n<rowBufferSize;++n)
+            {
+                if (rowBuffer[rowPos])
+                {
+                    traceOutQueue.enqueue(rowBuffer[rowPos]);
+                    rowBuffer[rowPos].clear();
+                }
+                if (++rowPos==rowBufferSize) rowPos = 0;
+            }
+        }
+        // Dump items to MemoryBuffer (without clearing items from queue)
+        void dump(MemoryBuffer &mb, IHThorArg *helper) const
+        {
+            unsigned rowPos = rowTailp;
+            CommonXmlWriter xmlwrite(XWFnoindent|XWFtrim);
+            IOutputMetaData *meta = helper->queryOutputMeta();
+            for (unsigned n=0; n<rowBufferSize;++n)
+            {
+                const CThorRowHistory &rowCurrent = rowBuffer[rowPos];
+                if (rowCurrent)
+                {
+                    xmlwrite.outputBeginNested("Row", true);
+                    xmlwrite.outputInt(rowCurrent.getSequence(), sizeof(int), "@seq");
+                    xmlwrite.outputInt(meta->getRecordSize(rowCurrent.get()), sizeof(size32_t), "@size");
+                    StringBuffer age;
+                    rowCurrent.getAgeInMs(age);
+                    xmlwrite.outputString(age.length(), age.str(), "@age");
+                    try
+                    {
+                        meta->toXML((const byte*)(rowCurrent.get()), xmlwrite);
+                    }
+                    catch (IException *e)
+                    {
+                        StringBuffer msg;
+                        e->errorMessage(msg);
+                        e->Release();
+                        DBGLOG("Row Tracing Error: %s", msg.str());
```

Comment:
minor: I think exception text and "Row" will be conjoined.
 Normally use EXCLOG(e, <message>); in this scenario.